### PR TITLE
Compose: Introduce an in-house `throttle()` utility, deprecate Lodash version

### PR DESCRIPTION
**Pull request by tyxla**
## What?
This PR introduces an in-house version of the `_.throttle()` function, along with some tests. I've borrowed most of the prior art by @sgomes (kudos for that), as indicated by his co-authorship of the first commit. We also update all `throttle` usages in Gutenberg to use the new version. Finally, we deprecate the original `_.throttle()` function so it will never be seen again in the repo.

## Why?

Lodash is known to unnecessarily inflate the bundle size of packages, and in most cases, it can be replaced with native language functionality. See these for more information and rationale:

* https://github.com/WordPress/gutenberg/issues/16938#issuecomment-602837246
* https://github.com/WordPress/gutenberg/issues/17025
* https://github.com/WordPress/gutenberg/issues/39495 

## How?

Similar to how. we introduced `debounce` in #43943, we're introducing `throttle` as another utility function in the `@wordpress/compose` package. That nicely gives us the opportunity to type it, which we're essentially doing. It comes with a few tests as well. We're exposing it externally since it will be used by other packages.

Furthermore, we update all `_.throttle()` calls to use the new utility function, including the `useThrottle()` hook. IMHO, that closes the loop of having an in-house hook for `useThrottle()` but using an external function for the actual throttling, giving us control on the types and functionality. 

You might ask why we're adding all the custom features (the `options` third argument) when they're not used internally. Well, because `useThrottle()` is exposed externally, and it was supporting all those options, so we have to continue supporting them for backwards compatibility. 

## Testing Instructions
* Verify all checks are green. 
* Smoke test all the editors.
* RN: Do some thorough smoke testing.
* Verify the places that use the new `throttle()` continue to work:
  * Drag and drop of blocks inside block areas
  * Drag and drop of images and videos inside block areas
  * Testing instructions from #33395

### CODE REVIEW

File: packages/compose/README.md [hunk 1]
```diff
@@ -161,6 +161,39 @@ _Related_
 Given a component returns the enhanced component augmented with a component
 only re-rendering when its props/state change
 
+### throttle
+
+A simplified and properly typed version of lodash's `throttle`, that
+always uses timers instead of sometimes using rAF.
+
+Creates a throttled function that only invokes `func` at most once per
+every `wait` milliseconds (or once per browser frame). The throttled function
```
Comment (sgomes) [inline packages/compose/README.md]: Similarly to the `debounce` PR, we should remove the mention of browser frames here. Looks like I missed it in the original code.

File: packages/compose/src/utils/throttle/index.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	leading?: boolean;
+	trailing?: boolean;
+}
+
+/**
+ * A simplified and properly typed version of lodash's `throttle`, that
+ * always uses timers instead of sometimes using rAF.
+ *
+ * Creates a throttled function that only invokes `func` at most once per
+ * every `wait` milliseconds (or once per browser frame). The throttled function
```
Comment (sgomes) [inline packages/compose/src/utils/throttle/index.ts]: Similarly to the `debounce` PR, we should remove the mention of browser frames here. Looks like I missed it in the original code.

**Review by sgomes — APPROVED** [review]
The code LGTM, thank you @tyxla! 👍 And thank you again for the co-authorship 🙂

File: packages/compose/README.md (hunk 1 above)
Comment (tyxla) [reply packages/compose/README.md]: Indeed, addressed in https://github.com/WordPress/gutenberg/pull/44225/commits/4ccb6206f475c69f6acb52f509e8e17210074ce3

File: packages/compose/src/utils/throttle/index.ts (hunk 2 above)
Comment (tyxla) [reply packages/compose/src/utils/throttle/index.ts]: Indeed, addressed in https://github.com/WordPress/gutenberg/pull/44225/commits/4ccb6206f475c69f6acb52f509e8e17210074ce3

**Pull request merged**